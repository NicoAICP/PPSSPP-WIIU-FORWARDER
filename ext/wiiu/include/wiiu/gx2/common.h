#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
