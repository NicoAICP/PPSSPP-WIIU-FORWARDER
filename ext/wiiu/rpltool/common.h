#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
