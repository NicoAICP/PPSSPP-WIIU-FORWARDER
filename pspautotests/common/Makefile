TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
