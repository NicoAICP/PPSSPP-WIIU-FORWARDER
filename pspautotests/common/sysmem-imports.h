extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();