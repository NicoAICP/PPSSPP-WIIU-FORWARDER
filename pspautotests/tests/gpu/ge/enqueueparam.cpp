#include <common.h>
#include <pspge.h>
#include <pspthreadman.h>
#include <psputils.h>

#include "../commands/commands.h"

extern "C" {
#include "sysmem-imports.h"
}

struct SceGeStack {
	unsigned int stack[8];
};

struct PspGeListArgs2 {
	unsigned int size;
	PspGeContext *context;
	u32 numStacks;
	SceGeStack *stacks;
};

static u32 __attribute__((aligned(16))) list[131072];

void testListParam(const char *title, PspGeListArgs2 *args, bool log = true) {
	PspGeContext ctx;
	int listID = sceGeListEnQueue(list, list + 1, -1, (PspGeListArgs *)args);
	if (listID >= 0) {
		if (log) {
			checkpoint("%s: OK", title);
		}
	} else {
		checkpoint("%s: %08x", title, listID);
	}
	sceGeBreak(1, NULL);
}

void testListParamHead(const char *title, PspGeListArgs2 *args, bool log = true) {
	PspGeContext ctx;
	int listID = sceGeListEnQueueHead(list, list + 1, -1, (PspGeListArgs *)args);
	sceGeContinue();
	if (listID >= 0) {
		if (log) {
			checkpoint("%s: OK", title);
		}
	} else {
		checkpoint("%s (Head): %08x", title, listID);
	}
	sceGeBreak(1, NULL);
}

extern "C" int main(int argc, char *argv[]) {
	memset(list, 0, sizeof(list));
	sceKernelDcacheWritebackRange(list, sizeof(list));

	PspGeListArgs2 args = { 0 };
	PspGeContext ctx;

	checkpointNext("Normal:");
	memset(&args, 0, sizeof(args));
	args.size = sizeof(args);
	testListParam("  Normal", &args);
	testListParamHead("  Normal", &args);

	checkpointNext("Sizes:");
	memset(&args, 0, sizeof(args));
	static const int sizes[] = { -1, 0, 1, 4, 5, 8, 12, 15, 16, 17, 32, 64, 1024 };
	for (int i = 0; i < ARRAY_SIZE(sizes); ++i) {
		args.size = sizes[i];
		char temp[256];
		snprintf(temp, sizeof(temp), "  Size=%d", sizes[i]);
		testListParam(temp, &args);
		testListParamHead(temp, &args);
	}

	checkpointNext("Context:");
	memset(&args, 0, sizeof(args));
	static const int ctx_sizes[] = { -1, 0, 4, 8, 12, 16 };
	for (int i = 0; i < ARRAY_SIZE(ctx_sizes); ++i) {
		memset(&ctx, 0xCC, sizeof(ctx));
		args.size = ctx_sizes[i];
		args.context = &ctx;
		testListParam("  With context", &args, false);
		checkpoint("  Context size=%d: %s", ctx_sizes[i], ctx.context[26] == 0xCCCCCCCC ? "not used" : "used");

		memset(&ctx, 0xCC, sizeof(ctx));
		testListParamHead("  With context", &args, false);
		checkpoint("  Context size=%d (Head): %s", ctx_sizes[i], ctx.context[26] == 0xCCCCCCCC ? "not used" : "used");
	}

	checkpointNext("Stack depths:");
	memset(&args, 0, sizeof(args));
	static const int stack_depths[] = { -1, 0, 1, 4, 5, 255, 256, 257, 1024 };
	for (int i = 0; i < ARRAY_SIZE(stack_depths); ++i) {
		args.size = sizeof(args);
		// Note that stacks is NULL in this case.
		args.numStacks = stack_depths[i];
		char temp[256];
		snprintf(temp, sizeof(temp), "  Stack depth=%d", stack_depths[i]);
		testListParam(temp, &args);
		testListParamHead(temp, &args);
	}

	checkpointNext("Stack depth by size:");
	static const int stack_sizes[] = { -1, 0, 4, 8, 12, 15, 16, 17, 32 };
	memset(&args, 0, sizeof(args));
	for (int i = 0; i < ARRAY_SIZE(stack_sizes); ++i) {
		args.size = stack_sizes[i];
		args.numStacks = 256;
		char temp[256];
		snprintf(temp, sizeof(temp), "  Stack=256 size=%d", stack_sizes[i]);
		testListParam(temp, &args);
		testListParamHead(temp, &args);
	}

	return 0;
}