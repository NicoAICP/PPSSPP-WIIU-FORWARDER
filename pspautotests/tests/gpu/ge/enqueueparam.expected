[x] Normal:
[x]   Normal: OK
[x]   Normal: OK

[x] Sizes:
[x]   Size=-1: OK
[x]   Size=-1: OK
[x]   Size=0: OK
[x]   Size=0: OK
[x]   Size=1: OK
[x]   Size=1: OK
[x]   Size=4: OK
[x]   Size=4: OK
[x]   Size=5: OK
[x]   Size=5: OK
[x]   Size=8: OK
[x]   Size=8: OK
[x]   Size=12: OK
[x]   Size=12: OK
[x]   Size=15: OK
[x]   Size=15: OK
[x]   Size=16: OK
[x]   Size=16: OK
[x]   Size=17: OK
[x]   Size=17: OK
[x]   Size=32: OK
[x]   Size=32: OK
[x]   Size=64: OK
[x]   Size=64: OK
[x]   Size=1024: OK
[x]   Size=1024: OK

[x] Context:
[x]   Context size=-1: used
[x]   Context size=-1 (Head): used
[x]   Context size=0: used
[x]   Context size=0 (Head): used
[x]   Context size=4: used
[x]   Context size=4 (Head): used
[x]   Context size=8: used
[x]   Context size=8 (Head): used
[x]   Context size=12: used
[x]   Context size=12 (Head): used
[x]   Context size=16: used
[x]   Context size=16 (Head): used

[x] Stack depths:
[x]   Stack depth=-1: 80000104
[x]   Stack depth=-1 (Head): 80000104
[x]   Stack depth=0: OK
[x]   Stack depth=0: OK
[x]   Stack depth=1: OK
[x]   Stack depth=1: OK
[x]   Stack depth=4: OK
[x]   Stack depth=4: OK
[x]   Stack depth=5: OK
[x]   Stack depth=5: OK
[x]   Stack depth=255: OK
[x]   Stack depth=255: OK
[x]   Stack depth=256: 80000104
[x]   Stack depth=256 (Head): 80000104
[x]   Stack depth=257: 80000104
[x]   Stack depth=257 (Head): 80000104
[x]   Stack depth=1024: 80000104
[x]   Stack depth=1024 (Head): 80000104

[x] Stack depth by size:
[x]   Stack=256 size=-1: 80000104
[x]   Stack=256 size=-1 (Head): 80000104
[x]   Stack=256 size=0: OK
[x]   Stack=256 size=0: OK
[x]   Stack=256 size=4: OK
[x]   Stack=256 size=4: OK
[x]   Stack=256 size=8: OK
[x]   Stack=256 size=8: OK
[x]   Stack=256 size=12: OK
[x]   Stack=256 size=12: OK
[x]   Stack=256 size=15: OK
[x]   Stack=256 size=15: OK
[x]   Stack=256 size=16: 80000104
[x]   Stack=256 size=16 (Head): 80000104
[x]   Stack=256 size=17: 80000104
[x]   Stack=256 size=17 (Head): 80000104
[x]   Stack=256 size=32: 80000104
[x]   Stack=256 size=32 (Head): 80000104
